contracts/token/ERC1155/ERC1155.sol

```diff
@@ -193,6 +198,24 @@ contract ERC1155 is ERC165, IERC1155
         _doSafeBatchTransferAcceptanceCheck(msg.sender, from, to, ids, values, data);
     }
 
+    /**
+     * @dev Register a token ID so other contract functionality knows this token
+     * actually exists and this ID is valid. Minting will automatically call this.
+     * @param id uint256 ID of the token to register
+     */
+    function _registerToken(uint256 id) internal virtual {
+        _tokenExists[id] = true;
```

Comment:
With [EIP2200](https://eips.ethereum.org/EIPS/eip-2200), an `SSTORE` here will cost the same as an `SLOAD` if `_tokenExists[id]` is already set to `true`, so this is already optimal in that sense.

We may want to consider using a 32-byte word instead of a boolean however: see the comments [here](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2171#pullrequestreview-394923232) (tl;dr: storing booleans is more costly than full-words because of extra checks placed by the compiler).

Reply:
Also, from the 1155 EIP:
>To broadcast the existence of a token ID with no initial balance, the contract SHOULD emit the TransferSingle event from 0x0 to 0x0, with the token creator as _operator, and a _value of 0.

What do you think about emitting such an event here? We should only do it the first time though, so we'd want to introduce an `if` statement anyway, even if not required for efficiency reasons.

Reply:
Ah, that is an interesting piece from the spec, I missed that one! I think it definitely would be a good idea to emit that one.
Also, I didn't even think of the double-registering case, thanks for your thoughts on that.
So, should I change `_tokenExists` to a `uint256` (using 0 and 1 for false and true) instead? I find it somewhat sad that a `bool` can end up being more expensive than a 32-byte type...

Reply:
Yes, I'm also not thrilled by that - we run into the same situation [on this other PR](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2171#issuecomment-615316756).

What do you think about using [`EnumerableSet`](https://docs.openzeppelin.com/contracts/3.x/api/utils#EnumerableSet) to store this data, instead of a plain mapping? As mentioned in the other comments, being able to enumerate this data is both important and hard to do via events. It'd mimic 721's [`tokenByIndex`](https://docs.openzeppelin.com/contracts/3.x/api/token/erc721#IERC721Enumerable-tokenByIndex-uint256-).

Reply:
Well, an `EnumerableSet` mostly makes sense for the reverse indexing, which from what I see is completely unneded here and only costs additional gas. Is there any reason for having this really enumerable? Also, it would not change anything wrt `bool` vs. `uint` as `EnumerableSet` only works on `bytes32` values, potentially in the variant of `UintSet` where they are cast to `uint256` anyhow.